### ISSUE TITLE
Enhancing Status Column on Shipments Dashboard [DISCUSSION]

### ISSUE THREAD

**Issue by hverdonk** [vsoch/freegenes#126]
@Koeng101 requested we discuss this issue over GitHub, so that our thoughts would be recorded. No action from @vsoch is required at the moment.

Currently, the Status column in the Shipments dashboard doesn't handle the case where we choose to reject someone's request for material. It also doesn't explain what we're

Based on Drew's comments, I propose we have two status columns. One will be the "Request Status," and will cover how far **we** have processed a request. There can be five possible values: 

1. Rejected - We have received a request, and chosen not to fulfill it. A simple toggle switch in the table pertaining to that request should suffice.
2. Awaiting Countersign - We have chosen to fulfill a request, and are waiting for the provided implementing letter to be countersigned. This will be the default status of new shipments.
3. Generating Label - The countersigned implementing letter has been uploaded. Now we have to create a shipment label. Triggered when the original implementation letter attached to a request is overwritten by a new, countersigned implementation letter.
4. Waiting to Ship - A shipment label has been generated, but the package itself has not been provided to FedEx. This status is set once the label has been generated.
4. Complete - The package has been shipped. FedEx takes over. The status would be set to "complete" once the request has been marked shipped via the existing button.
<img width="1126" alt="Screen Shot 2019-11-08 at 2 29 30 PM" src="https://user-images.githubusercontent.com/31907124/68515051-8324bf80-0234-11ea-94b0-477143472297.png">

Once FedEx takes over, we can track the package's shipping status in a "Shipment Status" column. The status in this column would be pulled from Shippo using their [tracking API](https://goshippo.com/docs/tracking/). Hovering over the status would display a mouseover with the substatus or definition such as those found [here](https://goshippo.com/docs/tracking/#event-definitions).

**Comment by hverdonk** [comment]
@Koeng101 it won't let me assign you to this issue, but I trust you can see it anyway.

**Comment by Koeng101** [comment]
TL;DR: Have two different shipping columns. Thoughts on this @vsoch

**Comment by vsoch** [comment]
Yes, it would be logical to have a "Completed" column that goes between Processing and Completed, and then an (actual) status column with some set of labels that map to the state of processing or completed. I would say something like:

 - Rejected: maps to Completed
 - Awaiting Countersign, Generating Label, Waiting to Ship: map to Processing
 - Shipped: (instead of complete?) of course maps to Complete

we would want to refactor the current status field to have the specific statuses, and then add a function to the model to `<instance>.get_status()` and determine processing / completed based on that.

**Comment by hverdonk** [comment]
Would it be possible to have a separate column tracking the shipment (i.e., FedEx) status of the package?

**Comment by vsoch** [comment]
Where does that information come from? If it's Shippo, I don't have great hope, haha. But if there is a function / api endpoint field you have in mind, it would definitely be do-able. My concern is that every time you load the table you would be pinging the shippo API. If it's not reliable / consistent, that's a hugely buggy dependency to add.

Keep in mind that the table has a fixed width, so we want to be selective about the columns we choose. If it gets too crowded it would be good to review which columns aren't absolutely essential for the main table (and could go on the orders detail page).

**Comment by hverdonk** [comment]
Shippo was the tracking API I was referring to, yes. There's some documentation about it [here](https://goshippo.com/docs/tracking/). I'm sure FedEx has one as well, though it makes for cleaner code to stick with Shippo at this point, especially since we can't guarantee that anyone else who builds this would also use FedEx. 

The tracking data wouldn't need to be loaded all that frequently - maybe querying the API every few hours or so, and only updating the table if the status had changed. Mostly, I just want an easy way to ensure that packages don't get stuck in customs without us knowing about it. Could you do some very basic exploration of the Shippo endpoint to determine if it's viable?

The columns themselves can be reviewed. I don't think we'd need both a "completed" column and an actual request status column - just the latter would be fine. We could also exchange the MTA column with the package tracking column (putting the information on the order details page), since users would have to upload an MTA to submit a request anyway. @Koeng101, thoughts?

**Comment by vsoch** [comment]
I'll take a look more closely tomorrow - if we can get some kind of update of a status based on an order, it's very reasonable to add a task to run every few hours to check, or on demand when clicking a button.

**Comment by vsoch** [comment]
This is a fairly large request so I'm going to move the FedEx tracking column to another issue - this one is scoped to be about the order statuses.

**Comment by vsoch** [comment]
See https://github.com/vsoch/freegenes/pull/129. I'm going to wait on implementing anything for #128 but I'm going to start to look into it and will post notes in #128. Once #129 is merged, then we can come back to dependency issues #127 and (not a dependency, but maybe a merge conflict! #128 )

**Comment by vsoch** [comment]
This is done.
